File: test/devices/core/registry/device_classification_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:openvitals/devices/core/registry/device_classification.dart';
import 'package:openvitals/devices/garmin/garmin_device_classifier.dart';
import 'package:openvitals/devices/wearos/wearos_device_classifier.dart';
import 'package:openvitals/domain/model/ble_sensor_models.dart';

void main() {
  const classifiers = [GarminDeviceClassifier(), WearosDeviceClassifier()];

  BleDiscoveredDevice device({
    String? name,
    bool advertisesSyncService = false,
  }) =>
      BleDiscoveredDevice(
        address: 'AA:BB:CC:DD:EE:FF',
        name: name,
        rssi: -50,
        suggestedCapabilities: const {},
        advertisesSyncService: advertisesSyncService,
      );

  DeviceClassification classify(BleDiscoveredDevice d) =>
      classifyDevice(d, classifiers);

  test('a Garmin watch product name → (garmin, watch)', () {
    final c = classify(device(name: 'vívoactive 5'));
    expect(c.integration, DeviceIntegration.garmin);
    expect(c.kind, BleDeviceKind.watch);
  });

  test('the member service alone does NOT make an unknown name a watch', () {
    // 0xFE1F surfaces a device in the scan, but the NAME decides the kind — an
    // unrecognised name stays a plain sensor rather than being swept up.
    final c = classify(device(name: 'anything', advertisesSyncService: true));
    expect(c.integration, isNull);
    expect(c.kind, BleDeviceKind.sensor);
  });

  test('a Garmin Edge name → (garmin, bikeComputer)', () {
    final c = classify(device(name: 'Edge 840'));
    expect(c.integration, DeviceIntegration.garmin);
    expect(c.kind, BleDeviceKind.bikeComputer);
  });

  test('a prefixed Edge name → (garmin, bikeComputer)', () {
    final c = classify(device(name: 'Garmin Edge 1040'));
    expect(c.integration, DeviceIntegration.garmin);
    expect(c.kind, BleDeviceKind.bikeComputer);
  });

  test('a member-service-only advert (no distinguishing name) → sensor', () {
    // Without a recognised name there is nothing to classify, so it falls
    // through to a plain live sensor rather than being assumed a watch.
    final c = classify(device(advertisesSyncService: true));
    expect(c.integration, isNull);
    expect(c.kind, BleDeviceKind.sensor);
  });

  test('a WearOS smartwatch name → (wearos, watch)', () {
    final c = classify(device(name: 'Galaxy Watch8 (89FZ)'));
    expect(c.integration, DeviceIntegration.wearos);
    expect(c.kind, BleDeviceKind.watch);
  });

  test('a live heart-rate strap → a plain sensor', () {
    final c = classify(device(name: 'TICKR'));
    expect(c.integration, isNull);
    expect(c.kind, BleDeviceKind.sensor);
    expect(c, same(DeviceClassification.sensor));
  });

  test('the NAME decides, so a WearOS name is WearOS even with 0xFE1F', () {
    // Contrived: a device advertising the Garmin member service but named like a
    // WearOS watch. Classification is name-based now, so it is WearOS — the
    // member service does not override a recognised name.
    final c =
        classify(device(name: 'Galaxy Watch', advertisesSyncService: true));
    expect(c.integration, DeviceIntegration.wearos);
  });

  test('an unnamed, unremarkable device → sensor', () {
    expect(classify(device()).integration, isNull);
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/devices/core/DeviceClassificationTest.kt
package tech.mmarca.openvitals.devices.core

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test
import tech.mmarca.openvitals.devices.garmin.GarminDeviceClassifier
import tech.mmarca.openvitals.devices.wearos.WearOsDeviceClassifier
import tech.mmarca.openvitals.domain.model.BleDeviceKind
import tech.mmarca.openvitals.domain.model.BleDiscoveredDevice
import tech.mmarca.openvitals.domain.model.DeviceIntegration

class DeviceClassificationTest {

    private val classifiers = listOf(GarminDeviceClassifier(), WearOsDeviceClassifier())

    @Test
    fun `a Garmin watch family classifies as a Garmin watch`() {
        val verdict = classifyDevice(discovered("vívoactive 5"), classifiers)
        assertEquals(DeviceIntegration.GARMIN, verdict.integration)
        assertEquals(BleDeviceKind.WATCH, verdict.kind)
    }

    @Test
    fun `an Edge classifies as a Garmin bike computer, not a watch`() {
        val verdict = classifyDevice(discovered("Edge 840"), classifiers)
        assertEquals(DeviceIntegration.GARMIN, verdict.integration)
        assertEquals(BleDeviceKind.BIKE_COMPUTER, verdict.kind)
    }

    @Test
    fun `a prefixed Edge name → (garmin, bikeComputer)`() {
        val verdict = classifyDevice(discovered("Garmin Edge 1040"), classifiers)
        assertEquals(DeviceIntegration.GARMIN, verdict.integration)
        assertEquals(BleDeviceKind.BIKE_COMPUTER, verdict.kind)
    }

    @Test
    fun `the member service alone does NOT make an unknown name a watch`() {
        // 0xFE1F surfaces a device in the scan, but the name decides the kind.
        val verdict = classifyDevice(
            discovered(name = "anything", advertisesSyncService = true),
            classifiers,
        )
        assertNull(verdict.integration)
        assertEquals(BleDeviceKind.SENSOR, verdict.kind)
    }

    @Test
    fun `a member-service-only advert (no distinguishing name) → sensor`() {
        // Without a recognised name it falls through to a plain live sensor.
        val verdict = classifyDevice(discovered(advertisesSyncService = true), classifiers)
        assertNull(verdict.integration)
        assertEquals(BleDeviceKind.SENSOR, verdict.kind)
    }

    @Test
    fun `the NAME decides, so a WearOS name is WearOS even with 0xFE1F`() {
        // A device advertising the Garmin member service but named like a WearOS watch is WearOS.
        val verdict = classifyDevice(
            discovered(name = "Galaxy Watch", advertisesSyncService = true),
            classifiers,
        )
        assertEquals(DeviceIntegration.WEAROS, verdict.integration)
    }

    @Test
    fun `an unnamed, unremarkable device → sensor`() {
        assertNull(classifyDevice(discovered(), classifiers).integration)
    }

    @Test
    fun `a wearos-style smartwatch classifies as a wearos watch`() {
        val verdict = classifyDevice(discovered("Galaxy Watch8 (89FZ)"), classifiers)
        assertEquals(DeviceIntegration.WEAROS, verdict.integration)
        assertEquals(BleDeviceKind.WATCH, verdict.kind)
    }

    @Test
    fun `anything unclaimed falls through to a plain sensor`() {
        val verdict = classifyDevice(discovered("Polar H10"), classifiers)
        assertEquals(DeviceClassification.SENSOR, verdict)
        assertNull(verdict.integration)
        assertEquals(BleDeviceKind.SENSOR, verdict.kind)
    }

    @Test
    fun `an HRM strap stays a sensor even though it is Garmin`() {
        val verdict = classifyDevice(discovered("HRM 200"), classifiers)
        assertEquals(DeviceClassification.SENSOR, verdict)
    }

    @Test
    fun `order matters, Garmin's verdict beats the generic watch name match`() {
        // A name that could match both classifiers: the first in the list wins.
        val verdict = classifyDevice(discovered("Garmin Forerunner 265S"), classifiers)
        assertEquals(DeviceIntegration.GARMIN, verdict.integration)
    }

    private fun discovered(
        name: String? = null,
        advertisesSyncService: Boolean = false,
    ): BleDiscoveredDevice = BleDiscoveredDevice(
        address = "AA:BB:CC:DD:EE:01",
        name = name,
        rssi = -60,
        suggestedCapabilities = emptySet(),
        advertisesSyncService = advertisesSyncService,
    )
}
